gapfill/gapfill.py

```diff
@@ -1,22 +1,227 @@
+        data_dir=data_dir
+    )
+    if budget_date_ranges_path is not None:
+        with Path(budget_date_ranges_path).open() as f:
+            budget_date_ranges = json.load(f)
+    else:
+        budget_date_ranges = {}
+
+    for model_dir in model_dirs:
+        model_dir = Path(model_dir)
+        site_data_dir = model_dir.parent.parent.parent
+
+        predictor_subset = model_dir.name
+        model = model_dir.parent.name
+        site = site_data_dir.name
+
+        gap_dir = site_data_dir / "gapfilled"
+        gap_dir.mkdir(exist_ok=True)
+
+        gap_path = site_data_dir / 'gap.csv'
+        gap_df = pd.read_csv(gap_path)
+        
+        model_obj = EnsembleModel(model_dir)
+
+        scale_path = model_dir / "scale.json"
+        if not scale_path.exists():
+            raise ValueError(
+                "Must run <python main.py test> with --distribution " +
+                f"{distribution} to compute an uncertainty scale before " +
+                "gapfilling."
+            )
+        with scale_path.open() as f:
+            uncertainty_scale = json.load(f)
+
+        y_hat_individual = model_obj.predict_individual(gap_df)
+        y_hat_mean = np.mean(y_hat_individual, axis=0)
+        y_hat_spread = (
+            (y_hat_individual - y_hat_mean) * uncertainty_scale + y_hat_mean
+        )
+        y_hat_dist = model_obj.predict_dist(
+            gap_df,
+            distribution=distribution,
+            uncertainty_scale=uncertainty_scale
+        )
+        y_hat_uncertainty_95 = get_pred_interval(y_hat_dist)
+
+        num_models = len(y_hat_individual)
+        gapfill_columns = {
+            'FCH4_F': y_hat_mean,
+            'FCH4_F_UNCERTAINTY': y_hat_uncertainty_95[:, 1] - y_hat_mean,
+            **{
+                f"FCH4_F{i+1}": y_hat_spread[i]
+                for i in range(num_models)
+            }
+        }
+        for column, gapfill_data in gapfill_columns.items():
+            gap_df[column] = gapfill_data
+
+        # Merge with original data
+        raw_path = site_data_dir / 'raw.csv'
+        raw_df = pd.read_csv(raw_path)
+
+        # Drop existing columns from the raw df
+        existing_fch4_columns = [
+            column for column in raw_df.columns
+            if 'FCH4_' in column
+        ]
+        raw_df.drop(columns=existing_fch4_columns, inplace=True)
+        gap_df = raw_df.merge(
+            gap_df[list(gapfill_columns.keys()) + ['TIMESTAMP_END']],
+            on='TIMESTAMP_END',
+            how='left'
+        )
+
+        # Add in observed values to FCH4_F, FCH4_F{1-N}
+        fch4_columns = [f'FCH4_F{i+1}' for i in range(num_models)]
+        observed_rows = gap_df['FCH4_F'].isna()
+        for fch4_column in ['FCH4_F'] + fch4_columns:
+            gap_df.loc[observed_rows, fch4_column] = (
+                gap_df.loc[observed_rows, 'FCH4']
+            )
+        # TODO: Should we keep uncertainty around observed values?
```

Comment:
Thoughts @gavinmcnicol ? Right now I'm setting them to 0. This is just for the data that's written to file, not the budget calculation.

Reply:
I think setting the uncertainty values around observed values to zero is fine, if you are referring to those in the new column(s) we create for the gap-filling uncertainties. We state in the text that the random uncertainties were not a part of our method. Though, for the FLUXNET-CH4 datasets, Sara Knox's uncertainty column will include them.

Reply:
Got it. Will keep this then, unless you think it would be better to remove the uncertainty column altogether.

Reply:
I think we can keep the uncertainty column if it does include the gap-filling uncertainty values.